### PR TITLE
Basic jemalloc command for printing arenas info with bin 

### PR DESCRIPTION
**Pull request by jetchirag**
For #2174

First command for jemalloc to print basic arenas details, it is not modularized and currently only for basic testing purpose.

```
pwndbg> arenas_info
Number of arenas:  1
Arena Address:  0x7ffff78010c0
Index   Address Slabcur
0 0x7ffff7814528 0x7ffff7816980
1 0x7ffff7814608 0x0
2 0x7ffff78146e8 0x0
...
```

### CODE REVIEW

**Comment by CptGibbon** [comment]
Awesome!
I'll get jemalloc installed and take a look at this 👍

**Comment by CptGibbon** [comment]
@jetchirag 
Huh I'm seeing the below error when I try the new command whilst debugging the example program from #2174 
What am I doing wrong?

Screenshot 2024-05-24 at 09 47 34

My jemalloc version is 5.3.0-182-gf9c0b5f7f8a917661db39289e38ec94d9d198f11

**Comment by jetchirag** [comment]
@CptGibbon Are you on dev branch? It was renamed, I'm using [5.3.0](https://github.com/jemalloc/jemalloc/releases/tag/5.3.0). Btw to update on progress, I'm currently working on parsing rtree which stores extent metadata, will add a command for that soon.

**Comment by CptGibbon** [comment]
> Are you on dev branch?

I'm on the `jemalloc_1` branch that this PR is based on.

EDIT: Oh I guess you meant dev branch of jemalloc?
In which case: yes.
Which branch should I use? I see `master`, `stable-3` & `stable-4`, but they're all considered "stale" with their last update at least a couple of years ago.

**Comment by jetchirag** [comment]
> > Are you on dev branch?
> 
> I'm on the `jemalloc_1` branch that this PR is based on.
> 
> EDIT: Oh I guess you meant dev branch of jemalloc? In which case: yes. Which branch should I use? I see `master`, `stable-3` & `stable-4`, but they're all considered "stale" with their last update at least a couple of years ago.

The latest release, version 5.3.0.

Edit: you can download it from https://github.com/jemalloc/jemalloc/releases/tag/5.3.0

**Comment by CptGibbon** [comment]
Ah okay thanks 👍
Screenshot 2024-05-24 at 14 25 52

**Comment by CptGibbon** [comment]
Great stuff, just a couple of notes for this PR:

1) I think renaming the command to something with "jemalloc" in the title makes it clear which allocator it's for (this will be cumbersome for now but if we implement automatic allocator detection later it could change to something more terse)
2) Specify somewhere which version of jemalloc this is for (perhaps just print it in the command output for now, we can do version detection later) that way people won't be confused like I was if it doesn't work
3) Amend `setup-dev.sh` so that reviewers don't have to manually install jemalloc and build your test program

**Comment by jetchirag** [comment]
@CptGibbon Added prefixes to command, and also installation script. Also pushed some WIP code for rtree parsing.

**Comment by CptGibbon** [comment]
> @CptGibbon Added prefixes to command, and also installation script. Also pushed some WIP code for rtree parsing.

Okay thanks, great work 👍
I would recommend creating tests for these new jemalloc commands before adding any more to this PR.
Let me know if you want help writing them.

**Comment by CptGibbon** [comment]
I'm seeing a few lint commits, I find working in an IDE that supports devcontainers can alleviate this.
pwndbg has a devcontainer setup that will automatically install the appropriate formatting & linting extensions and run them each time you save a file.

**Comment by jetchirag** [comment]
So I have removed the old commands and added proper class for generating memory mapping by jemalloc. There are lots of todos, I'll be adding one or two classes for rtree (including one to finding extent information from memory pointer address given by `malloc()`) and other structures then move onto actual command followed by tests as I'd also like to get started with test early since there's that discussed problem of testing memory addresses.

Things were slow past 2 weeks, I did read a lot of documentations on the same and will try to pick up pace now.

Also, thank you for dev containers suggestion though seems like it didn't fix linting on `setup-dev.sh` or maybe I forgot to run linting on it. It's just whitespace, I'll fix that in next commit, hopefully

**Comment by jetchirag** [comment]
@CptGibbon I think now we have something which resembles `malloc_chunk` and `heap` command which is the primary goal. The structure's still not there yet and tests aren't done but it would be great if I can get feedback on this so we can make changes early.

**Comment by CptGibbon** [comment]
> @CptGibbon ... it would be great if I can get feedback on this so we can make changes early.

No problem, I'll review today.
Thanks for the update 🙏

**Comment by CptGibbon** [comment]
@jetchirag
I loaded your test program into GDB and ran your new commands.
Screens of my experience below:

### jemalloc_heap
Screenshot 2024-06-20 at 15 56 46

`jemalloc_heap` appears to work, though some of those addresses look spurious, can you confirm this is correct behavior?

### find_extent
Screenshot 2024-06-20 at 16 02 27

`find_extent` appears to work, again could you confirm this is correct behavior?

### extent_info
Used by the `jemalloc_heap` and `find_extent` commands.

### jemalloc_base_info
Screenshot 2024-06-20 at 16 06 04

No dice.

### get_extent
Screenshot 2024-06-20 at 16 08 48

Not sure if I'm using the correct address as an argument here but nothing I try works...

### jemalloc_test
Screenshot 2024-06-20 at 16 09 45

One last note; if you want your test program to get linked with jemalloc you'll need to add a corresponding line to the makefile that builds tests.

**Comment by jetchirag** [comment]
Thank you for testing @CptGibbon 
Currently only `jemalloc_heap`,`jemalloc_find_extent`and `jemalloc_extent_info` are to be used, rest are just for testing purpose. Since I've changed the structure a lot, some old test ones would have broke, it's fine.

Your output for heap and find extent seems to be expected. Yes there are some false addresses, some of it seems to be due to cache_oblivious, I'll investigate that and add note in output if needed or fix it. Any comment on the code itself?

**Comment by CptGibbon** [comment]
> Any comment on the code itself?

I'm usually of the mind that if it works it works, but I can take a closer look 👍

So far my only concern aside from the address issue is that `jemalloc_heap` was rather slow to complete, at around 3 seconds. This was running on a free codespaces instance of the smallest variety, but that's still rather slow compared to most commands.

Great work 👌

**Comment by jetchirag** [comment]
> So far my only concern aside from the address issue is that jemalloc_heap was rather slow to complete, at around 3 seconds. This was running on a free codespaces instance of the smallest variety, but that's still rather slow compared to most commands.

Yes that has been my concern as well. I've mentioned this in comment, it's happening due to the size of rtree. I was trying to find it there's an alternate way, perhaps it maintains caches or thread specific information, will check.

**Comment by CptGibbon** [comment]
Just 1 note and 1 question:

A current development goal is to uncouple as much code from the gdb library as possible, hopefully this makes testing easier and makes room for different debugger backends.

One step towards this is `gdblib.memory.fetch_struct_as_dictionary()`
https://github.com/pwndbg/pwndbg/blob/f21000356f819f5074e9a10ed74dedda89243da5/pwndbg/gdblib/memory.py#L388-L397

Which will eventually become something like `debugger.fetch_struct_as_dictionary()`. 

An example use can be found in `gdblib.heap.ptmalloc.fetch_chunk_metadata()`.

The transition to debugger-agnostic code is still underway, so if using the above function would slow you down too much it's not a show-stopper, but it's something to consider.

I see you credit jegdb in the source, which code specifically is taken from jegdb?

**Comment by jetchirag** [comment]
@CptGibbon I'll check that function. I don't think there's much code taken directly from that script. The code for bit shifting bitfield was directly used 

https://github.com/pwndbg/pwndbg/pull/2176/files#diff-c96fafbf3d980a07d7f741d99185b64e1a09dd7371ac617dc6f7bffb25b7f68bR380-R389

and `rtree_leaf_elm_bits_extent_get`

Primarily I used the script as base to correlate with jemalloc source code to understand structure.

**Comment by jetchirag** [comment]
> One last note; if you want your test program to get linked with jemalloc you'll need to add a corresponding line to the makefile that builds tests.

Is it fine to directly link jemalloc path as such in makefile:
```
heap_jemalloc_extent_info.out: heap_jemalloc_extent_info.c
	@echo "[+] Building heap_jemalloc_extent_info.out"
	${CC} -g -O0 -o heap_jemalloc_extent_info.out heap_jemalloc_extent_info.c -pthread -lpthread /usr/local/lib/libjemalloc.a -lm -lstdc++ -pthread
```

**Comment by CptGibbon** [comment]
> Is it fine to directly link jemalloc path as such in makefile:

Yes that seems to be the current project standard.

@disconnect3d are you happy with this approach?

**Comment by jetchirag** [comment]
I noticed the duplicate flag in makefile, will remove that

**Comment by jetchirag** [comment]
I looked at cache stores in thread specific information structure and elsewhere but we won't be able to use it for finding extents. I think I will have to stick to current method for now, unless someone has a better approach.

I'll utilize the function you linked to. I'll have to compile jemalloc with cache oblivious disabled to check those weird addresses in heap command.

**Comment by jetchirag** [comment]
Can you please also answer if these commands (just the functionality) fulfil the primary scope of this project?

**Comment by CptGibbon** [comment]
> I'll utilize the function you linked to.

Great!

> I'll have to compile jemalloc with cache oblivious disabled to check those weird addresses in heap command.

Okay thanks, let us know if that changes anything.

> Can you please also answer if these commands (just the functionality) fulfil the primary scope of this project?

I think these fulfill the jemalloc portion of the project 👍
Once we've ironed out the existing issues we can ask the maintainers to merge it and you're welcome to move onto the kernel portion of the project.

Thanks for your hard work so far 🙏

**Comment by jetchirag** [comment]
@CptGibbon mypy is not sure about the type of sub dicts in GdbDict from that struct as dictionary function and thus there are several error messages in linting. Is there any preferred way to deal with this?

```
pwndbg/gdblib/heap/jemalloc.py:383: error: Item "dict[str, GdbDict | int]" of "GdbDict | int" has no attribute "address"  [union-attr]
                    bins_addr = int(self._Value["bins"].address)
```

**Comment by CptGibbon** [comment]
> @CptGibbon mypy is not sure about the type of sub dicts in GdbDict from that struct as dictionary function and thus there are several error messages in linting

That sub dict is a @gsingh93 creation so we'd best ask them.

File: .devcontainer/devcontainer.json
Comment (disconnect3d) [inline .devcontainer/devcontainer.json]: Why do we need this?

File: pwndbg/commands/heap.py
Comment (disconnect3d) [inline pwndbg/commands/heap.py]: 5.3.0 or newer too? or we dont know for sure?

Maybe: "(This command was tested only for jemalloc 5.3.0)"?

File: pwndbg/commands/heap.py
Comment (disconnect3d) [inline pwndbg/commands/heap.py:1578]: > Memory is conceptually broken into extents. Extents are always aligned to multiples of the page size. This alignment makes it possible to find metadata for user objects quickly. User objects are broken into two categories according to size: small and large. Contiguous small objects comprise a slab, which resides within a single extent, whereas large objects each have their own extents backing them.

hmm... wonder if we can at least add a comment on those extents somewhere (aka what are they lol) or inform the user about them, idk...

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: Please change to a link to github (or other VCS) with specific version this is based from (5.3.0), or even commit

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: What is relevant place?

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: The arch name is 'aarch64' or 'arm64'. But anyway, can we make this work for x86-64 as well?

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: Do we know what does SC stand for? Can we add a comment for it?

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: ?

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: What does this mean? Shall we comment it?

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: What does this mean? Shall we comment it?

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: What does this mean? Shall we comment it?

**Comment by disconnect3d** [comment]
Hi, I added a few comments here and there. It would be nice to document as much "why" and "what" as possible for further readers or maintainers of the code, as otherwise, it may be extremely hard to take this work upon and extend it or adjust for other jemalloc versions or needs etc.

Ideally, this could also explain some implementation details of jemalloc - aka how it allocates memory, what are extents, design choices etc etc.

File: .devcontainer/devcontainer.json
Comment (jetchirag) [reply .devcontainer/devcontainer.json]: This was added to disable ASLR which was failing for me locally. I'll remove/git ignore it.

File: pwndbg/commands/heap.py
Comment (jetchirag) [reply pwndbg/commands/heap.py]: What about "This command was tested only for jemalloc 5.3.0 and does not support lower versions"? Since we know for sure lower versions are not supported due to different structure.

File: pwndbg/gdblib/heap/jemalloc.py
Comment (jetchirag) [reply pwndbg/gdblib/heap/jemalloc.py]: According to https://github.com/jemalloc/jemalloc/blob/5.3.0/include/jemalloc/internal/quantum.h#L21, value for it is 4 as well so we can remove this TODO.

File: pwndbg/gdblib/heap/jemalloc.py
Comment (jetchirag) [reply pwndbg/gdblib/heap/jemalloc.py]: Added

File: pwndbg/commands/heap.py
Comment (jetchirag) [reply pwndbg/commands/heap.py:1578]: Added information on extent in comments

**Comment by jetchirag** [comment]
@disconnect3d @CptGibbon I've added more comments describing most of the code and jemalloc itself.

**Comment by CptGibbon** [comment]
Thanks for the latest update and comments 🙏

Here's my experience using the new commands:

### jemalloc_heap

jemalloc_heap

How come each extent is printed twice?
I'm not seeing any spurious addresses any more though 👍
It would be great to have a "small class: True" example for the sake of tests.

### jemalloc_extent_info

jemalloc_extent_info

Looks good 👍
Any reason the version warning is different from the other two commands?

### jemalloc_find_extent

jemalloc_find_extent

Am I using this incorrectly? How come the extent address is off by 2? That didn't happen before...

I think this work is so nearly done, just a couple more things to square away.

**Comment by jetchirag** [comment]
@CptGibbon Different message was unintentional.

For the last issue, the allocated address is incorrect so there is some issue, it couldn't locate correct extent address. I'm checking why that happens.

**Comment by jetchirag** [comment]
@CptGibbon Can you please share the code file you ran to get these outputs? How much memory did you allocate?

**Comment by CptGibbon** [comment]
> @CptGibbon Can you please share the code file you ran to get these outputs? How much memory did you allocate?

I used your test files, I believe the output above is from `tests/gdb-tests/tests/binaries/heap_jemalloc_heap.c`

**Comment by jetchirag** [comment]
> How come each extent is printed twice?

Accidentally had it printing it twice in some cases. Fixed.


> It would be great to have a "small class: True" example for the sake of tests

Yes, it was there but missing in heap due to last changes. Fixed and now visible.

**Comment by jetchirag** [comment]
Sample outputs:

```
pwndbg> jemalloc_heap
Jemalloc heap
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Allocated Address: 0x7ffff7000000
Extent Address: 0x7ffff7a16a00
Size: 0x400000
Small class: False
State: Active

Allocated Address: 0x7ffff7800000
Extent Address: 0x7ffff7a16500
Size: 0x7000
Small class: False
State: Active

Allocated Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580
Size: 0x1000
Small class: True
State: Active

Allocated Address: 0x7ffff7808000
Extent Address: 0x7ffff7a16980
Size: 0x1f8000
Small class: False
State: Retained
```

```
pwndbg> jemalloc_find_extent ptr2
Jemalloc find extent
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Pointer Address: 0x7ffff7000000
Extent Address: 0x7ffff7a16a00

Allocated Address: 0x7ffff7000000
Extent Address: 0x7ffff7a16a00
Size: 0x400000
Small class: False
State: Active

pwndbg> jemalloc_find_extent ptr
Jemalloc find extent
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Pointer Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580

Allocated Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580
Size: 0x1000
Small class: True
State: Active
```

```
pwndbg> jemalloc_extent_info 0x7ffff7a16580
Jemalloc extent info
This command was tested only for jemalloc 5.3.0 and does not support lower versions

Allocated Address: 0x7ffff7807000
Extent Address: 0x7ffff7a16580
Size: 0x1000
Small class: True
State: Active
```

**Comment by CptGibbon** [comment]
@jetchirag thank you for the latest update

If you're happy with the state of this work, you can take it out of draft state and we'll give it a final review 👍

**Comment by jetchirag** [comment]
@CptGibbon Moved out of draft. There might be changes needed on structuring the code since everything is currently in same file. Let me know if any.

**Review by CptGibbon — APPROVED** [review]
I think this satisfies the GSoC criteria set out earlier this year.
It gives pwndbg a foot in the door of jemalloc inspection and I'm sure this feature will see plenty of use. I look forward to seeing future contributions to it by yourself & others.

There have been a lot of changes to how pwndbg fundamentally operates during this working period (my chance to encourage more continuous integration style contributions) so I defer to the honorable @disconnect3d for any pwndbg-specific feedback.

Thank you @jetchirag for your hard work and for continuing despite losing your laptop!

**Comment by jetchirag** [comment]
@CptGibbon Thank you, I'll remove merge conflicts in next commit.

File: .devcontainer/devcontainer.json
Comment (disconnect3d) [reply .devcontainer/devcontainer.json]: Please remove this flag. One can disable ASLR globally outside of the container if needed
```suggestion
```

File: pwndbg/gdblib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/gdblib/heap/jemalloc.py]: @jetchirag is there any way we can do  this elsewhere? I mean, we are excepting a huge block of code here :/

**Comment by disconnect3d** [comment]
Btw the commands are a bit slow, is there a way to cache it or make it faster? Why is it so slow?

**Comment by CptGibbon** [comment]
> Btw the commands are a bit slow, is there a way to cache it or make it faster? Why is it so slow?

I believe parsing the rtree is rather costly.
My two cents is that this feature works and performance wasn't part of the original spec, so perhaps we could let the author continue working on their kernel feature which is due shortly and then revisit this feature to discuss improvements.

**Comment by disconnect3d** [comment]
Commited a merge commit here to (hopefully) fix conflicts.

**Comment by disconnect3d** [comment]
@jetchirag Ping, do u maybe know why the tests fail on CI?

**Comment by disconnect3d** [comment]
Ok, the tests were failing bcoz this string was updated in code, but not in tests:

Old:
> This command only support jemalloc 5.3.0

New:
> This command was tested only for jemalloc 5.3.0 and does not support lower versions

I changed this in the tests now.

File: pwndbg/aglib/heap/jemalloc.py
Comment (disconnect3d) [inline pwndbg/aglib/heap/jemalloc.py:414]: What is this property for? It seems to be unused across the code and it seems to be broken: it uses `self._bins` which is not defined in the `Arena` class.

cc: @jetchirag

**Comment by disconnect3d** [comment]
Fixed mypy issues in jemalloc.py. Also, it seems that its `Arena` class along its `slabs` property are unused. Is this intended? Do we plan to extend it in the future or what? Shall we remove it?

cc: @jetchirag @CptGibbon

File: pwndbg/aglib/heap/jemalloc.py
Comment (jetchirag) [reply pwndbg/aglib/heap/jemalloc.py:414]: @disconnect3d I left this for future extension as I started working on this project with this property but it's not needed right now. I'll remove Arenas.

**Pull request merged**
